### PR TITLE
feat:add devel image support

### PR DESCRIPTION
**Pull request by czc-unac**
## What does this PR do?
Add support for additional devel image.

## Changes
- build_and_push_cann.yml add image-devel option.
- cann_publish_version.json add devel mapping.

### CODE REVIEW

File: .github/workflows/build_and_push_cann.yml [hunk 1]
```diff
@@ -242,7 +423,7 @@ jobs:
       matrix:
         include:
           - arch: x86_64
-            runner: linux-x86_64
+            runner: ubuntu-latest
```
Comment (wjunLu) [inline .github/workflows/build_and_push_cann.yml]: revert this

File: .github/workflows/build_and_push_cann.yml [hunk 2]
```diff
@@ -274,14 +455,25 @@ jobs:
           username: ${{ secrets.DOCKER_USERNAME }}
           password: ${{ secrets.DOCKER_TOKEN }}
 
+      - name: Set image repository
```
Comment (wjunLu) [inline .github/workflows/build_and_push_cann.yml]: The base and `devel` images are in the same repo, please delete this step

File: .github/workflows/build_and_push_cann.yml [hunk 3]
```diff
@@ -201,7 +382,7 @@ jobs:
       matrix:
         include:
           - arch: x86_64
-            runner: linux-x86_64
+            runner: ubuntu-latest
```
Comment (wjunLu) [inline .github/workflows/build_and_push_cann.yml]: revert, keep the online settings

File: .github/workflows/build_and_push_cann.yml (hunk 1 above)
Comment (czc-unac) [reply .github/workflows/build_and_push_cann.yml]: fixed

File: .github/workflows/build_and_push_cann.yml (hunk 2 above)
Comment (czc-unac) [reply .github/workflows/build_and_push_cann.yml]: deleted

File: .github/workflows/build_and_push_cann.yml (hunk 3 above)
Comment (czc-unac) [reply .github/workflows/build_and_push_cann.yml]: fixed

File: .github/workflows/build_and_push_cann.yml [hunk 4]
```diff
@@ -186,6 +186,187 @@ on:
           - 8.1.rc1-910b-openeuler22.03-py3.10
           - 8.1.rc1-910b-openeuler22.03-py3.11
           - 8.1.rc1-910b-openeuler24.03-py3.10
+          - 9.0.0-950-ubuntu22.04-py3.11-devel
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml]: Only add the images to be published.

File: cann_publish_version.json [hunk 5]
```diff
@@ -1078,11 +1078,1117 @@
           "8.1.rc1-910b-openeuler22.03-py3.11"
       ]
     },
-    {
-      "path": "cann/8.1.rc1-910b-openeuler24.03-py3.10",
-      "tags": [
-          "8.1.rc1-910b-openeuler24.03-py3.10"
-      ]
-    }
+{
```
Comment (wxsIcey) [inline cann_publish_version.json]: ditto

File: .github/workflows/build_and_push_cann.yml [hunk 6]
```diff
@@ -280,8 +461,8 @@ jobs:
         with: 
           platforms: ${{ matrix.arch }}
           tags: |
-            docker.io/ascendai/cann
-            quay.io/ascend/cann
+            ${{ steps.set-repo.outputs.docker_repo }}
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml]: Where the devel image to publish?

**Comment by wjunLu** [comment]
sign CLA

**Comment by czc-unac** [comment]
> sign CLA

signed

File: .github/workflows/build_and_push_cann.yml [hunk 7]
```diff
@@ -287,19 +287,37 @@ jobs:
           outputs: type=image,push-by-digest=true,name-canonical=true,push=true
           provenance: false
 
-      - name: Export digest
+      - name: Build and push devel by digest
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml]: I think your logits is wrong. You must build devel image after pushing base image. Do you have ever testing it?

File: tools/template/ubuntu.devel.Dockerfile.j2 [hunk 8]
```diff
@@ -0,0 +1,61 @@
+FROM swr.cn-south-1.myhuaweicloud.com/ascendhub/{{ item.tag }}
```
Comment (wxsIcey) [inline tools/template/ubuntu.devel.Dockerfile.j2]: how th distinct openeuler and ubuntu base image?

File: tools/template/ubuntu.devel.Dockerfile.j2 [hunk 9]
```diff
@@ -0,0 +1,60 @@
+FROM quay.io/ascend/cann:{{ item.cann_version }}-{{ item.cann_chip }}-{{ item.os_name }}{{ item.os_version }}-py{{ item.py_version }}
```
Comment (wxsIcey) [inline tools/template/ubuntu.devel.Dockerfile.j2:1]: if using os_version， will use the Dockerfille generate From xxx:xxx-openEuler-xxx tag

File: .com [hunk 10]
```diff
@@ -0,0 +1,327 @@
+
```
Comment (wxsIcey) [inline .com]: why add this file?

File: tools/template/ubuntu.devel.Dockerfile.j2 (hunk 9 above)
Comment (czc-unac) [reply tools/template/ubuntu.devel.Dockerfile.j2:1]: I see. I need to lower it.

**Comment by wxsIcey** [comment]
Please get my approval before merging this PR. I'm worried about potential issues afterward. Thank you.

**Review by wxsIcey — APPROVED** [review]

File: .github/workflows/build_and_push_cann.yml [hunk 11]
```diff
@@ -411,3 +417,363 @@ jobs:
...
+          skopeo copy --all \
+          docker://quay.io/ascend/cann:${{ inputs.cann_tag }}-devel \
+          docker://swr.cn-southwest-2.myhuaweicloud.com/base_image/ascend-ci/cann:${{ inputs.cann_tag }}-devel
+
+          skopeo copy --all \
+          docker://quay.io/ascend/cann:${{ inputs.cann_tag }}-devel \
+          docker://swr.cn-east-3.myhuaweicloud.com/ascendhub-test/cann:${{ inputs.cann_tag }}-devel
+
+
+  only-build-push-devel-digest:
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml:603]: consider merge `only-build-push-devel-digest` and `build-push-devel-digest`, maybe in next pr

File: .github/workflows/build_and_push_cann.yml [hunk 12]
```diff
@@ -411,3 +417,363 @@ jobs:
...
+            elif [[ "$TAG" == *"docker.io/ascendai/cann"* ]]; then
+              docker buildx imagetools create -t "$TAG"-devel \
+                $(printf 'docker.io/ascendai/cann@sha256:%s ' *)
+            else
+              echo "Error: Unknown registry in TAG: $TAG"
+              exit 1
+            fi
+          done
+
+  only-sync-image:
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml:745]: ditto

File: .github/workflows/build_and_push_cann.yml [hunk 13]
```diff
@@ -411,3 +417,363 @@ jobs:
...
+
+      - name: Upload digest
+        uses: actions/upload-artifact@v4
+        with:
+          name: digests-${{ inputs.cann_tag }}-devel-${{ matrix.arch }}
+          path: ${{ runner.temp }}/devel-digests/*
+          if-no-files-found: error
+          retention-days: 1
+  
+  only-push-devel-image:
```
Comment (wxsIcey) [inline .github/workflows/build_and_push_cann.yml:670]: ditto

**Comment by wxsIcey** [comment]
I will merge the pr first, please submit a follow pr to merge `only-build-push-devel-digest `and `build-push-devel-digest`.
